### PR TITLE
[graph partition] reorder to reduce #partitions for simple dependencies

### PR DESCRIPTION
**Pull request by BoyuanFeng**
This PR reduces #graph partitions by reordering nodes when the `should_partition` nodes have simple dependencies. Specifically, for `should_partition` nodes:
    a. If a node has no dependency or only depends on graph inputs: move to the front. Use case is when we move symints to cuda tensor for PaddedTensorSubclass
    b. If the only user of a node is OutputNode: move it to the end.

#### Example

The following example shows a padded tensor subclass use case where we copy symint to a cuda tensor (aka mask) in the middle of function. Reordering still generates 1 cudagraph by moving the mask to the front.

```python
import torch

torch._inductor.config.graph_partition = True

# Two reasons for this:
# 1. We want to reuse the same mask for many masked_fill calls
# 2. Prevent inductor from fusing this op into other ops (e.g. masked_fill)
#    so we can still reorder in scheduler
@torch.library.custom_op("mylib::create_mask", mutates_args=(), tags=(torch._C.Tag.cudagraph_unsafe,))
def create_mask(padded_size: int, original_size: int, device: torch.device) -> torch.Tensor:
    mask = torch.zeros((padded_size,), dtype=torch.bool, device=device)
    mask[original_size:] = True
    return mask

@create_mask.register_fake
def _(padded_size, original_size, device):
    return torch.empty((padded_size,), dtype=torch.bool, device=device)


def f(padded_tensor, original_tensor, weight):
    original_size = original_tensor.size()[0]
    padded_size = padded_tensor.size()[0]
    
    # element wise op so we don't care padding value
    padded_tensor = padded_tensor + 1
    padded_tensor = torch.nn.functional.relu(padded_tensor)

    # dot product requires padding with 0
    dot_res = padded_tensor.dot(weight)
    padded_tensor += dot_res

    # min requires padding with inf, so we create mask now
    mask = create_mask(padded_size, original_size, padded_tensor.device)
    min_res = torch.min(
        torch.ops.aten.masked_fill(padded_tensor, mask, float("inf"))
    )

    # max requires padding with inf. we can reuse previous mask
    max_res = torch.max(
        torch.ops.aten.masked_fill(padded_tensor, mask, -float("inf"))
    )

    return min_res+max_res+padded_tensor

compiled_f = torch.compile(f, mode="reduce-overhead")

def run(padded_size, original_size):
    padded_tensor = torch.randn(padded_size, device="cuda")
    padded_tensor[original_size:] = 0
    original_tensor = torch.randn(original_size, device="meta")

    weight = torch.randn(padded_size, device="cuda")
    eager_out = f(padded_tensor, original_tensor, weight)
    compiled_out = compiled_f(padded_tensor, original_tensor, weight)
    assert torch.allclose(eager_out[0], compiled_out[0])
    assert torch.allclose(eager_out[1], compiled_out[1])

# new cudagraph
run(8, 4)

# new cudagraph due to recompile
run(8, 6)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
...
+                    buffer := self.name_to_buf.get(dep.name, None)
+                ) and buffer.get_device() != node_device:
+                    input_device = buffer.get_device()
+                    # the dependent node may have been processed
+                    if input_device_nodes := device_to_nodes.pop(input_device, None):
+                        result.extend(input_device_nodes)
+            device_to_nodes[node_device].append(node)
+
+        for node in nodes:
+            visit(node)
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4245]: I dont follow this.. we visit each node, and in each node we add `device_to_nodes[node_device].append(node)`. what is result supposed to represent ?

we always add each node in line 4242. and then there are some nodes we add twice. then we add those nodes to result.

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
...
+        result: list[BaseSchedulerNode] = []
+        device_to_nodes: dict[Optional[torch.device], list[BaseSchedulerNode]] = (
+            collections.defaultdict(list)
+        )
+
+        def visit(node: BaseSchedulerNode) -> None:
+            node_device = node.get_device()
+            for dep in node.unmet_dependencies:
+                if (
+                    buffer := self.name_to_buf.get(dep.name, None)
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4236]: when would name_to_buf.get(dep.name) not exist ? can we more explicitly handle the non case

**Review by eellison — COMMENTED** [review]
have some thoughts. dont need to do the most complicated thing just yet, but worth at least gauging what parts of what we view as solution we want in short/medium/long term.

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
...
+        self, nodes: list[BaseSchedulerNode]
+    ) -> list[BaseSchedulerNode]:
+        """
+        Reorder a node if it should be partitioned and has simple dependency:
+        1. move a partitioned node to the front if it has no dependency
+        2. move a partitioned node to the back if it is only used by OutputNode
+        3. otherwise do not reorder
+        """
+
+        front: list[BaseSchedulerNode] = []
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/scheduler.py:4262]: Btw - the sum() thing is a real use case - this is basically what they're doing in vllm with wanting to make sure the quant on top of attention (downcast to fp8) is not in the cudagraph region..

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
...
+        middle: list[BaseSchedulerNode] = []
+        back: list[BaseSchedulerNode] = []
+
+        def only_output_user(node: BaseSchedulerNode) -> bool:
+            for buf in node.get_outputs():
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4267]: One other thing, semi relevant again to vllm.. 

If we have an extern op that is not cudagraphable in the middle of the graph and takes in an output, we can allocate the output inside the prior cudagraph region to avoid copy.

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -4269,6 +4276,38 @@ def get_graph_partition_signature(
...
+                for use in buf.users:
+                    if not isinstance(use.node, OutputNode):
+                        return False
+            return True
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/scheduler.py:4298]: yes i think it does.. WeakDep adds as a user

**Review by eellison — APPROVED** [review]
